### PR TITLE
Draft: Add new buffer implementation

### PR DESCRIPTION
**Pull request by Lowentwickler**
This is a WIP PR with renderer refactoring.

- [x] New buffer implementation
- [x] Make Pipeline layout and Mesh layout ready for Storage buffers

### CODE REVIEW

**Comment by Lowentwickler** [comment]
@QuantumEntangledAndy What do you think about the Buffer code?

**Comment by QuantumEntangledAndy** [comment]
Seems to be fine. Let see how it works in practice and go from there.

**Comment by Lowentwickler** [comment]
Good, I will continue in this manner. First re-implement existing parts and re-arrange them in the module, then solve the layout issue.
I want to avoid direct usage of all native WGPU components outside of `renderer` module, but keep this possibility for end-user for any kind of customization.

**Comment by QuantumEntangledAndy** [comment]
The use_as ones should be marked with `#[must_use]`

**Comment by Lowentwickler** [comment]
Yeah, of course, I will solve all CI stuff in the end.

**Comment by QuantumEntangledAndy** [comment]
That's fair having them open for dev to customise would be useful.

I am working on the voxel/terrain stuff at the moment. Will be adding `dotrix_voxel` which handles rendering/marching cubing arbitary voxels, and `dotrix_terrain` which generates and updates voxels for the terrain. Direct rendering of voxels is coming along ok and I'm trying thinking about how it will fit in with marching cubes too.

File: dotrix_core/src/renderer/pipelines.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    /// [`Id`] of the shader
+    pub shader: Id<Shader>,
+    /// Pipeline bindings
+    pub bindings: Bindings,
+    /// renderer's cycle
+    pub cycle: usize,
+    /// is disabled
+    pub disabled: bool,
+}
+
```
Comment (Lowentwickler) [inline dotrix_core/src/renderer/pipelines.rs:19]: Here are two separate components for Render and Compute pipelines.

File: dotrix_pbr/src/skeletal.rs [hunk 2]
```diff
@@ -108,16 +105,16 @@ pub fn render(
         &mut Pose,
         &mut Material,
         &mut Transform,
-        &mut Pipeline,
+        &mut Render,
     )>();
```
Comment (Lowentwickler) [inline dotrix_pbr/src/skeletal.rs:110]: Here is how render system has changed.

**Review by Xaniaro — APPROVED** [review]
I had a problem with differentiate `Render`, `Renderer`, `RenderOptions` structs. Now it's ok.
Good Work!

Btw what are string labels used for? For storage, or some bindings in shaders?
For example:
```rust
Buffer::uniform("Pose Buffer")
TextureBuffer::new("Texture")
TextureBuffer::new("CubeMap Texture Buffer")
```

**Review by QuantumEntangledAndy — APPROVED** [review]
I am ok with it I don't see why we can't :). LGTM

**Comment by Lowentwickler** [comment]
> I had a problem with differentiate `Render`, `Renderer`, `RenderOptions` structs

`Render` - is a Component, containing initial pipeline data with bindings, cycle counter and control(disable) flag.
`RenderOptions` - GPU specific options for the pipeline (like Depth Buffer usage) that are not supposed to be changed during the Pipeline lifetime
`Renderer` - is a service, as it was before.



> Btw what are string labels used for? For storage, or some bindings in shaders?

This labels are being passed straight to WGPU. If something goes wrong, WGPU mentions this labels in debug output and you know where the problem is.

**Comment by Xaniaro** [comment]
Great. Thanks for explanation 👍

**Pull request merged**